services.py

```diff
@@ -85,45 +91,83 @@ def get_param_from_config(self, config_path: str, param_name: str):
-            try:
-                os.remove(path)
-                self.logger.info(f'Deleted file: {path}')
-            except PermissionError:
-                self.logger.error(f'Error deleting file: {path}')
-
-
-    def get_logger(self, name):
-        log_format = '%(asctime)s  %(name)8s  %(levelname)5s  %(message)s'
-        logging.basicConfig(level=logging.DEBUG,
-                            format=log_format,
-                            filename='dev.log',
-                            filemode='w')
-        return logging.getLogger(name)
-
-
-    def get_city_from_coordinates(self, latitude, longitude):
-        geoLoc = Nominatim(user_agent="GetLoc")
-
-        loc = self.get_lat_lon()
-        locname = geoLoc.reverse(f'50.0 40.0')
-        address = locname.raw['address']
-        print(address)
-
-    # latitude, longitude = get_current_location()
-    # if latitude and longitude:
-    #     city = get_city_from_coordinates(latitude, longitude)
-    #     if city:
-    #         print(f"The current city is: {city}")
-    #     else:
-    #         print("Could not determine the city.")
-    # else:
-    #     print("Could not determine the current location.")
+        try:
+                gps = serial.Serial(SERIAL_PORT, baudrate = 9600, timeout = 0.5)
+                
+                data = gps.readline().decode('utf-8').rstrip()
+                message = data[0:6]
+                # print(message)
+                if (message == "$GNRMC"):
+                    # GPRMC = Рекомендуемые минимальные конкретные данные GPS / транзит
+                    # Чтение фиксированных данных GPS является альтернативным подходом, 
+                    # который также работает
+                    parts = data.split(",")
+                    print(parts)
+                    if parts[2] == 'V':
+                        # V = Предупреждение, скорее всего, спутников нет в поле зрения ...
+                        print ("GPS receiver warning")
+                    else:
+                        # Получить данные о местоположении, которые были переданы с сообщением GPRMC. 
+                        # В этом примере интересуют только долгота и широта. Для других значений можно
+                        # обратиться к http://aprs.gids.nl/nmea/#rmc
+                        longitude = self.formatDegreesMinutes(parts[5], 3)
+                        latitude = self.formatDegreesMinutes(parts[3], 2)
+                        logger.info("Device position: lon = " + str(longitude) + ", lat = " + str(latitude))
+                else:
+                    # Обрабатывать другие сообщения NMEA и неподдерживаемые строки
+                    pass
+
+                gps.close()
+                return [latitude, longitude]
+        except Exception as e:
+            logger.error(f'Error getting GPS coordinates through serial port: {e}\n Falling back to IP geolocation...')
+            g = geocoder.ip('me')
+            return g.latlng
```

Comment:
Pretty sure that `geocoder` can also raise exceptions. These lines should be in their own `try`/`except` block, and not in the `except:` portion of anything.

Reply:
👍Should have thought of this.